proxy/api/src/ethereum/address.rs

```diff
@@ -0,0 +1,46 @@
+use serde::{Deserialize, Serialize};
+use std::{
+    convert::{TryFrom, TryInto},
+    str::FromStr,
+};
+
+/// An [EIP-55](https://eips.ethereum.org/EIPS/eip-55) Ethereum address.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(try_from = "String", into = "String")]
+pub struct Address(String);
+
+#[derive(Clone, Debug, thiserror::Error)]
+#[error("invalid EIP-55 address")]
+pub struct InvalidEIP55AddressError;
+
+impl TryFrom<String> for Address {
+    type Error = InvalidEIP55AddressError;
+
+    fn try_from(address: String) -> Result<Self, Self::Error> {
+        if eip55::validate_address(&address) {
+            Ok(Self(address))
+        } else {
+            Err(InvalidEIP55AddressError)
+        }
+    }
+}
+
+impl FromStr for Address {
+    type Err = InvalidEIP55AddressError;
+
+    fn from_str(address: &str) -> Result<Self, Self::Err> {
+        address.to_string().try_into()
```

Comment:
nit: since `validate_address` takes a `&str` it would indicate that `FromStr` should be the implementation and `TryFrom<String>` can use the `FromStr` impl. That means we're not sticking a `&str` on the heap every time we parse.

Reply:
Nitpick time? Nitpick time! I was thinking about 2 cases, constructing `Address` from `&str` and from `String`.

In case of `&str` we need to copy it onto heap as a `String` at some point anyway. In the current implementation it's a wasted effort only if an error occurs.

In case of `String` we don't need to copy anything, just wrap it in an `Address`. If we push a `String` through `from_str`, it's effectively cloned and wasting resources every time.

Reply:
Ah good point! Nitpick time over!